app/platform/vfs.c

```diff
@@ -0,0 +1,469 @@
+    dir_level++;
+    if (level = c_strchr( level, '/' )) {
+      level++;
+    } else {
+      break;
+    }
+  }
+#endif
+
+#ifdef BUILD_SPIFFS
+  if (fs_fns = myspiffs_realm( normpath, &outname, TRUE )) {
+    // our SPIFFS integration doesn't support directories
+    if (c_strlen( outname ) == 0) {
+      ok = VFS_RES_OK;
+    }
+  }
+#endif
+
+#ifdef BUILD_FATFS
+  if (fs_fns = myfatfs_realm( normpath, &outname, TRUE )) {
+    if (c_strchr( outname, ':' )) {
+      // need to set FatFS' default drive
+      fs_fns->chdrive( outname );
+      // and force chdir to root in case path points only to root
+      fs_fns->chdir( "/" );
+    }
+    if (fs_fns->chdir( outname ) == VFS_RES_OK) {
+      ok = VFS_RES_OK;
+    }
+    c_free( outname );
+  }
+#endif
+
+  return ok == VFS_RES_OK ? VFS_RES_OK : VFS_RES_ERR;
+}
+
+sint32_t vfs_errno( const char *name )
+{
+  vfs_fs_fns *fs_fns;
+  const char *normname = normalize_path( name );
+  char *outname;
+
+  if (!name) name = "";  // current drive
+
+#ifdef BUILD_SPIFFS
+  if (fs_fns = myspiffs_realm( normname, &outname, FALSE )) {
+    return fs_fns->ferrno( );
+  }
+#endif
+
+#ifdef BUILD_FATFS
+  if (fs_fns = myfatfs_realm( normname, &outname, FALSE )) {
+    sint32_t r = fs_fns->ferrno( );
+    c_free( outname );
+    return r;
+  }
+#endif
+
+  return VFS_RES_ERR;
+}
+
+sint32_t vfs_ferrno( int fd )
+{
+  vfs_file *f = (vfs_file *)fd;
+
+  if (f) {
+    return f->fns->ferrno ? f->fns->ferrno( f ) : 0;
+  } else {
+    vfs_fs_fns *fs_fns;
+    const char *name = "";  // current drive
+    char *outname;
+
+#ifdef BUILD_SPIFFS
+    if (fs_fns = myspiffs_realm( name, &outname, FALSE )) {
+      return fs_fns->ferrno( );
+    }
+#endif
+
+#ifdef BUILD_FATFS
+    if (fs_fns = myfatfs_realm( name, &outname, FALSE )) {
+      sint32_t r = fs_fns->ferrno( );
+      c_free( outname );
+      return r;
+    }
+#endif
+  }
+}
+
+
+void vfs_clearerr( const char *name )
+{
+  vfs_fs_fns *fs_fns;
+  char *outname;
+
+  if (!name) name = "";  // current drive
+
+  const char *normname = normalize_path( name );
+
+#ifdef BUILD_SPIFFS
+  if (fs_fns = myspiffs_realm( normname, &outname, FALSE )) {
+    fs_fns->clearerr( );
+  }
+#endif
+
+#ifdef BUILD_FATFS
+  if (fs_fns = myfatfs_realm( normname, &outname, FALSE )) {
+    fs_fns->clearerr( );
+    c_free( outname );
+  }
+#endif
+}
+
+const char *vfs_basename( const char *path )
+{
+  const char *basename;
+
+  // deduce basename (incl. extension) for length check
+  if (basename = c_strrchr( path, '/' )) {
+    basename++;
+  } else if (basename = c_strrchr( path, ':' )) {
+    basename++;
+  } else {
+    basename = path;
+  }
+
+  return basename;
+}
+
+
+// ---------------------------------------------------------------------------
+// supplementary functions
+//
+int vfs_getc( int fd )
+{
+  unsigned char c = 0xFF;
+  sint32_t res;
+
+  if(!vfs_eof( fd )) {
+    if (1 != vfs_read( fd, &c, 1 )) {
+      NODE_DBG("getc errno %i\n", vfs_ferrno( fd ));
```

Comment:
On of the cloud builds just failed because of this line:

```
vfs.c: In function 'vfs_getc':
vfs.c:457:7: error: implicit declaration of function 'c_printf' [-Werror=implicit-function-declaration]
       NODE_DBG("getc errno %i\n", vfs_ferrno( fd ));
       ^
cc1: all warnings being treated as errors
```

The build configuration was as follows:

> branch=dev
> modules=enduser_setup,file,gpio,http,mqtt,net,node,tmr,uart,wifi
> ssl-enabled=false
> debug-enabled=true